### PR TITLE
feat(inbound-filters): Add react-hydration-errors filter

### LINKED ISSUE #45038: New Inbound Filter: React Hydration Errors
### Problem Statement

About this ticket: https://github.com/getsentry/sentry/issues/44877

We re-evaluated our decision and decided for now that we are going to, by default, ignore hydration error as a new Inbound Filter. 
The reason for this is, currently, those errors are not actionable, we can't reference the component that triggered it, therefore, it's hard to understand and you usually just want to ignore it. 
Maybe in the future we revisit this if we are able to detect the root cause in the SDK to make this actionable but for now, ignoring them will result in a better user experience.

### Solution Brainstorm

Add a new Inbound Filter:
`Filter out hydration errors`
`React falls back to do a full re-render on a page, these errors are often not actionable.`
image

We should make this, retroactively a default for all projects and when enabled, Relay filters those errors out.

Here are the error messages we want to filter: 
```
  'https://reactjs.org/docs/error-decoder.html?invariant=419', // The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.
  'https://reactjs.org/docs/error-decoder.html?invariant=422', // There was an error while hydrating this Suspense boundary. Switched to client rendering.
  'https://reactjs.org/docs/error-decoder.html?invariant=423', // There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root...
  'https://reactjs.org/docs/error-decoder.html?invariant=425'  // Text content does not match server-rendered HTML...
  ```
ref: https://github.com/getsentry/sentry-javascript/issues/6295#issuecomment-1352958925

### PR DESCRIPTION
**Pull request by priscilawebdev**
**Problem:**

We want to make it clear how Sentry is transforming error payloads, specifically for react-minified errors. This is because it can be hard to tell how to filter these errors (using either beforeSend or server-side inbound filters) as the UI shows the transformed message, not the original one.

**Solution:**

A new "Inbound Filters" called "Filter out hydration errors" is added via this PR, so users only have to toggle and relay will filter out the most common react hydration errors.

**More Details:**
Fixes: https://github.com/getsentry/sentry/issues/45038
Fixes: https://github.com/getsentry/sentry-javascript/issues/6295

### CODE REVIEW

File: src/sentry/ingest/test.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{"event_id":"65fad5091a3f4db1b9812777f6dcc052","sent_at":"2023-02-28T16:05:58.058Z","sdk":{"name":"sentry.javascript.nextjs","version":"7.39.0"}}
```
Comment (andriisoldatenko) [inline src/sentry/ingest/test.json:1]: why do we need this file?

Comment (RaduW) [reply src/sentry/ingest/test.json:1]: Yep... I think it was included by mistake @priscilawebdev  ?

**Review by RaduW — APPROVED** [review]
With the exception of the Json file (which I think was added by mistake) everything seems fine.
Please remove the `test.json` (unless I completely miss something) then I think it's good to go.

File: src/sentry/ingest/test.json (hunk 1 above)
Comment (priscilawebdev) [reply src/sentry/ingest/test.json:1]: yes that was included by mistake... I am removing and turning this back to "ready to review"  😉

File: tests/sentry/api/endpoints/snapshots/ProjectFiltersTest/test_get.pysnap [hunk 2]
```diff
@@ -14,6 +14,12 @@ source: tests/sentry/api/endpoints/test_project_filters.py
   hello: localhost - Filter out events coming from localhost
   id: localhost
   name: Filter out events coming from localhost
+- active: false
```
Comment (priscilawebdev) [inline tests/sentry/api/endpoints/snapshots/ProjectFiltersTest/test_get.pysnap:17]: @andriisoldatenko @RaduW I used the following command to update this file 😉

`SENTRY_SNAPSHOTS_WRITEBACK=1 pytest tests/sentry/api/endpoints/test_project_filters.py`

File: src/sentry/ingest/inbound_filters.py [hunk 3]
```diff
@@ -231,3 +234,9 @@ class _LegacyBrowserFilterSerializer(serializers.Serializer):
     description="Some crawlers may execute pages in incompatible ways which then cause errors that"
     " are unlikely to be seen by a normal user.",
 )
+
+_react_hydration_errors_filter = _FilterSpec(
+    id=FilterStatKeys.REACT_HYDRATION_ERRORS,
+    name="Filter out hydration errors",
+    description="React falls back to do a full re-render on a page and these errors are often not actionable.",
```
Comment (iker-barriocanal) [inline src/sentry/ingest/inbound_filters.py:241]: Are we targeting all hydration errors, or only these coming from React?

File: src/sentry/relay/config/__init__.py [hunk 4]
```diff
@@ -115,6 +115,23 @@ def get_filter_settings(project: Project) -> Mapping[str, Any]:
...
+                # The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.
+                "https://reactjs.org/docs/error-decoder.html?invariant=419",
+                # There was an error while hydrating this Suspense boundary. Switched to client rendering.
+                "https://reactjs.org/docs/error-decoder.html?invariant=422",
+                # There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.
+                "https://reactjs.org/docs/error-decoder.html?invariant=423",
+                # Text content does not match server-rendered HTML.
+                "https://reactjs.org/docs/error-decoder.html?invariant=425",
+            ]
+            error_messages = [*error_messages, *react_known_hydration_errors]
```
Comment (iker-barriocanal) [inline src/sentry/relay/config/__init__.py:133]: `error_messages` here may be `None` if the option above is not registered, and in that case the unpacking fails.

File: src/sentry/relay/config/__init__.py [hunk 5]
```diff
@@ -115,6 +115,23 @@ def get_filter_settings(project: Project) -> Mapping[str, Any]:
...
+
         if error_messages:
             filter_settings["errorMessages"] = {"patterns": error_messages}
```
Comment (iker-barriocanal) [inline src/sentry/relay/config/__init__.py:136]: `errorMessages` are [here](https://github.com/getsentry/relay/blob/b710ffdaa1eb0f95e5a0cd5fa80ee10fc4d73e5a/relay-filter/src/config.rs#L200) in relay, and `patterns` are [globs](https://github.com/getsentry/relay/blob/b710ffdaa1eb0f95e5a0cd5fa80ee10fc4d73e5a/relay-filter/src/config.rs#L136). Do you want to escape special characters in the links above, like `?`?

**Review by iker-barriocanal — COMMENTED** [review]
(I've only looked at the relay part.) Does this come with a PR in relay?

**Comment by priscilawebdev** [comment]
Sorry @iker-barriocanal but I just completely changed the logic.

File: src/sentry/api/serializers/models/project.py [hunk 6]
```diff
@@ -176,6 +176,7 @@ def format_options(attrs: defaultdict(dict)):
             "sentry:performance_issue_creation_rate"
         ),
         "filters:blacklisted_ips": "\n".join(options.get("sentry:blacklisted_ips", [])),
+        "filters:react-hydration-errors": bool(options.get("filters:react-hydration-errors", True)),
```
Comment (priscilawebdev) [inline src/sentry/api/serializers/models/project.py:179]: @jan-auer testing locally, even if I send "True" this option never gets updated. do you know why? maybe I am missing something

**Review by jan-auer — APPROVED** [review]

**Comment by priscilawebdev** [comment]
frontend https://github.com/getsentry/sentry/pull/45300

File: src/sentry/relay/config/__init__.py [hunk 7]
```diff
@@ -109,14 +109,31 @@ def get_filter_settings(project: Project) -> Mapping[str, Any]:
...
+            # Hydration failed because the initial UI does not match what was rendered on the server.
+            "https://reactjs.org/docs/error-decoder.html?invariant=418",
+            # The server could not finish this Suspense boundary, likely due to an error during server rendering. Switched to client rendering.
+            "https://reactjs.org/docs/error-decoder.html?invariant=419",
+            # There was an error while hydrating this Suspense boundary. Switched to client rendering.
+            "https://reactjs.org/docs/error-decoder.html?invariant=422",
+            # There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.
+            "https://reactjs.org/docs/error-decoder.html?invariant=423",
+            # Text content does not match server-rendered HTML.
+            "https://reactjs.org/docs/error-decoder.html?invariant=425",
```
Comment (jjbayer) [inline src/sentry/relay/config/__init__.py:132]: Because `error_messages` parses glob patterns, we could shorten this to

```suggestion
            "https://reactjs.org/docs/error-decoder.html?invariant={418,419,422,423,425}"
```

Comment (HazAT) [reply src/sentry/relay/config/__init__.py:132]: Let's please leave it like this, obviously more verbose, but the comments are helpful to understand what the number actually means.

Comment (jjbayer) [reply src/sentry/relay/config/__init__.py:132]: I'm totally in favor of keeping the comments, but we could still shorten the value itself.

Comment (priscilawebdev) [reply src/sentry/relay/config/__init__.py]: ok, I have updated the code shortening the value but still keeping the comments... Hope it's better now
